### PR TITLE
[LLVMGPU] Refactor VirtualMMAIntrinsic to it's own attribute and enum.

### PR DESCRIPTION
**Pull request by raikonenfnu**
VirtualMMA are variants of MMAOps that has layouts modified through interleaved of reads and/or unrolled-K. These are especially useful for coalescing reads from shared memory to register and for FA to align layouts between the 1st and 2nd matmul.

However, the current implementation of virtual MMA lives in MMAAttr and MMAIntrinsic which is a violation of abstractions. Especially since data_tiled_mma_layout is dependent on it. Additionally, having VirtualMMA in MMAAttr also limits the flexibility of adding more stuff and specification to it in the future.

In order to resolve this conflict of abstractions, we are migrating VirtualMMAAttr to it's own enums and attr/class.

Things that we have changed in this PR:

1. Factor our VMFMA from `MMAAttr` and `MMAIntrinsic` into it's own attr and enums `VirtualMMAAttr ` and `VirtualMMAIntrinsic`.
2. Update places that may use `MMAAttr` or `VirtualMMAAttr` to simply use `MmaInterfaceAttr` such as `KernelConfig`, `LLVMGPUConfigureTensorLayout`, and `AMDGPUDistributeContract`, `GPUNestedDistribution`
3. Move `get[A,B,C]SingleSubgroupLayout` as an global method S.T we can cast as `MmaInterfaceAttr` in places that may use `MMAAttr` and `VirtualMMAAttr`.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td [hunk 1]
```diff
@@ -60,6 +60,48 @@ def IREEGPU_MmaInterfaceAttr : AttrInterface<"MmaInterfaceAttr"> {
       /*methodName=*/"getABCVectorTypes",
       /*args=*/(ins)
     >,
+    InterfaceMethod<
```
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td]: For every `InterfaceMethod` added in any new PR, I ask: why can't this simply be a global function declared in a `.h` instead? Two concerns here:
1. The C++ debate between member and non-member functions. It's been an evolution in the whole C++ culture (some googling found [this post](https://www.reddit.com/r/cpp/comments/262it3/comment/chn3vhw/) which summarizes a few links) and in the C++ standard library itself (e.g. `vector.begin()` vs `std::begin(vector)`)
2. In this instance it's also about going through generated code vs not. With `InterfaceMethod`, the indirection through generated code makes it harder to follow the code, both for the human reader, and even for tools like `clangd` which can't tell how many stack frames to walk to when you click "go to implementation".

In this instance, these were existing global functions, so why change that now?

**Review by bjacob — COMMENTED** [review]
Thanks for taking this on! +1 on the general approach. Question below about changing normal functions to `InterfaceMethod`'s.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 2]
```diff
@@ -1179,6 +1115,262 @@ FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
...
+  }
+  case VirtualMMAIntrinsic::VMFMA_F32_32x32x16_F16: {
+    auto aType = VectorType::get({8}, A);
+    auto bType = VectorType::get({8}, B);
+    auto cType = VectorType::get({16}, C);
+    return std::make_tuple(aType, bType, cType);
+  }
+  }
+  // This should not happen but just to make GCC happy.
+  return std::make_tuple(VectorType{}, VectorType{}, VectorType{});
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]:
```suggestion
  return {};
```

Also add `llvm_ureachable` if it's not supposed to happen?

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 3]
```diff
@@ -1179,6 +1115,262 @@ FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
...
+  // amd_matrix_instruction_calculator tells us about the number of 32-bit
+  // registers. So need to adjust accordingly. All vectors should be 1-D.
+  auto [A, B, C] = getABCElementTypes();
+  switch (getIntrinsic().getValue()) {
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F8E4M3FNUZ:
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F16: {
+    auto aType = VectorType::get({8}, A);
+    auto bType = VectorType::get({8}, B);
+    auto cType = VectorType::get({4}, C);
+    return std::make_tuple(aType, bType, cType);
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]:
```suggestion
    return {aType, bType, cType};
```

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 4]
```diff
@@ -1179,6 +1115,262 @@ FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
...
+  }
+  case VirtualMMAIntrinsic::VMFMA_F32_32x32x16_F16: {
+    auto aType = VectorType::get({8}, A);
+    auto bType = VectorType::get({8}, B);
+    auto cType = VectorType::get({16}, C);
+    return std::make_tuple(aType, bType, cType);
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]:
```suggestion
    return {aType, bType, cType};
```

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 3 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: In c++17, I don't think there are any reasons to use `make_pair`/`make_tuple`/`make_optional`

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 5]
```diff
@@ -1179,6 +1115,262 @@ FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
...
+
+int64_t VirtualMMAAttr::getSubgroupSize() const {
+  switch (getIntrinsic().getValue()) {
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F8E4M3FNUZ:
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F16:
+  case VirtualMMAIntrinsic::VMFMA_F32_32x32x16_F16: {
+    return 64;
+  }
+  }
+  // This should not happen but just to make GCC happy.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp:1235]: also here

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 6]
```diff
@@ -283,6 +278,56 @@ def IREEGPU_DataTiledMMAAttr :
...
+    "populateOperandOffsetsSizesStrides",
+    "buildMmaOperation",
+  ]>
+]> {
+  let mnemonic = "virtual_mma_layout";
+  let cppNamespace = "::mlir::iree_compiler::IREE::GPU";
+
+  let description = [{
+    This mma variant represents "virtual" MMA ops that has modification to
+    it's native layouts by unrollK and/or interleave reads. The |intrinsic|
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]:
```suggestion
    its native layouts by unrollK and/or interleave reads. The |intrinsic|
```

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h [hunk 7]
```diff
@@ -14,6 +14,19 @@
...
 
+namespace mlir::iree_compiler::IREE::GPU {
+
+// Partial nested layout for an MMA intrinsic's matrix input/output inside
+// a single subgroup.
+struct MMASingleSubgroupLayout {
+  llvm::SmallVector<int64_t, 2> outer;
+  llvm::SmallVector<int64_t, 2> thread;
+  llvm::SmallVector<int64_t, 2> tstrides;
+  llvm::SmallVector<int64_t, 2> element;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h]:
```suggestion
  SmallVector<int64_t, 2> outer;
  SmallVector<int64_t, 2> thread;
  SmallVector<int64_t, 2> tstrides;
  SmallVector<int64_t, 2> element;
```

**Review by kuhar — COMMENTED** [review]
Looks good overall, just some nits

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 2 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: @kuhar , `llvm_unreachable` is a powerful optimization tool, by allowing the compiler to assume a place to be unreachable, but it's also dangerous if it's ever reached, so it only makes sense to use it in very, very performance-critical code.  In all other cases, if you want to catch something that's not supposed to happen, maybe you want an assertion (`assert(false)`); all the usual objections against assertions ("don't risk crashing the compiler on bad user input") also apply even more strongly against `llvm_unreachable`.

Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: Also note that here, GCC is saying something very important and Clang's behavior is borderline buggy: these TableGen-generated enums unfortunately have a specified underlying type:

```
enum class MMAIntrinsic : uint32_t {
```

Unfortunately, the `: uint32_t` makes a big difference here: it makes it legal to form values that aren't any of the enumerator values. So when Clang says "you dont need a default: case in this switch because it would be UB anyway if the enum value were none of the enumerators", that's only correct for enums without a specified underlying type.

Here, unless you want to audit every caller, we have to treat it as a possibility that the enum value is none of the enumerators and so that "error" case is in fact reachable.

Note: https://eel.is/c++draft/expr.static.cast#9 :
> If the enumeration type has a fixed underlying type, the value is first converted to that type by integral promotion ([[conv.prom]](https://eel.is/c++draft/conv.prom)) or integral conversion ([[conv.integral]](https://eel.is/c++draft/conv.integral)), if necessary, and then to the enumeration type[.](https://eel.is/c++draft/expr.static.cast#9.sentence-2)
If the enumeration type does not have a fixed underlying type, the value is unchanged if the original value is within the range of the enumeration values ([[dcl.enum]](https://eel.is/c++draft/dcl.enum)), and otherwise, the behavior is undefined[.](https://eel.is/c++draft/expr.static.cast#9.sentence-3)

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: The way I understand it, `llvm_unreachable` translates to:
* Debug ==> `assert(false)`
* Release ==> `assume(false)`

While plain `assert` disappears in Release. So to me `llvm_unreachable` is fine as a tool to stop the compiler from complaining about missing return values etc., when there's something else that structurally checks that all the cases were handled. And I think this will happen by warnings about unhandled `switch` cases.

But I don't care about it much in either direction

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: I replied to the first comment before I saw your second one; assert makes sense to me then. I didn't realize it was an open set of possible values.

Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: > While plain assert disappears in Release.

Yes. Apparently the idea is that the same control paths are exercised in Debug and Release builds. That assumption seems untenable in practice. Users use Release bugs, and they will run into corner cases that weren't hit by tests and local developer invocations on debug builds.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td (hunk 1 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td]: > In this instance, these were existing global functions, so why change that now?

Great question! Firstly, I think these method has always been a member method of MmaAttr, and previously we have  always just casted attributes into `MmaAttr` and then calling these `get[A,B,C]SingleSubgroupLayout`

e.g:
1. https://github.com/iree-org/iree/blob/2b1a8e7679fe0c7a9fcd076eb1c16b9d76cfc24b/compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp#L311-L318
2. https://github.com/iree-org/iree/blob/2b1a8e7679fe0c7a9fcd076eb1c16b9d76cfc24b/compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp#L1316 https://github.com/iree-org/iree/blob/2b1a8e7679fe0c7a9fcd076eb1c16b9d76cfc24b/compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp#L1491

Now that it can also be `VirtualMmaAttr` and since `VirtualMmaAttr` and `MmaAttr` does not share a base class, the path of least motion is to move these member methods of `MmaAttr` into `MmaInterfaceAttr`. 

That being said, I do agree it is probably safer and cleaner to move these methods as global method, so let me push that up now. :)

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 2 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: good catch! done!

EDIT: Meant this as review comment, please ignore for now.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 5 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp:1235]: done!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td (hunk 6 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: done

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 2 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: Filed https://github.com/llvm/llvm-project/issues/115345

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 3 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: done, thanks!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h (hunk 7 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.h]: done!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 8]
```diff
@@ -345,6 +336,40 @@ static ConcreteMmaLayout getConcreteMFMALayout(MLIRContext *context,
...
 }
 
+//===----------------------------------------------------------------------===//
+// MmaInterface Attribute Helper Functions
+//===----------------------------------------------------------------------===//
+
+MMASingleSubgroupLayout getASingleSubgroupLayout(MmaInterfaceAttr mmaKind) {
+  if (auto mmaAttr = llvm::dyn_cast<MMAAttr>(mmaKind)) {
+    return mmaAttr.getASingleSubgroupLayout();
+  } else if (auto vmmaAttr = llvm::dyn_cast<VirtualMMAAttr>(mmaKind)) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: nit: no else after return: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 9]
```diff
@@ -345,6 +336,40 @@ static ConcreteMmaLayout getConcreteMFMALayout(MLIRContext *context,
...
+    return vmmaAttr.getASingleSubgroupLayout();
+  }
+  assert(false && "unhandled MMA Interface type.");
+  return {};
+}
+
+MMASingleSubgroupLayout getBSingleSubgroupLayout(MmaInterfaceAttr mmaKind) {
+  if (auto mmaAttr = llvm::dyn_cast<MMAAttr>(mmaKind)) {
+    return mmaAttr.getBSingleSubgroupLayout();
+  } else if (auto vmmaAttr = llvm::dyn_cast<VirtualMMAAttr>(mmaKind)) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td (hunk 1 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td]: I think it's done, LMK if this is what you have in mind or if you think we can clean it up some more!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 10]
```diff
@@ -345,6 +336,40 @@ static ConcreteMmaLayout getConcreteMFMALayout(MLIRContext *context,
   return concreteLayout;
 }
 
+//===----------------------------------------------------------------------===//
+// MmaInterface Attribute Helper Functions
+//===----------------------------------------------------------------------===//
+
+MMASingleSubgroupLayout getASingleSubgroupLayout(MmaInterfaceAttr mmaKind) {
+  if (auto mmaAttr = llvm::dyn_cast<MMAAttr>(mmaKind)) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: nit: you shouldn't need `llvm::` in front of casts? Also everywhere else

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 11]
```diff
@@ -345,6 +336,40 @@ static ConcreteMmaLayout getConcreteMFMALayout(MLIRContext *context,
...
+    return vmmaAttr.getBSingleSubgroupLayout();
+  }
+  assert(false && "unhandled MMA Interface type.");
+  return {};
+}
+
+MMASingleSubgroupLayout getCSingleSubgroupLayout(MmaInterfaceAttr mmaKind) {
+  if (auto mmaAttr = llvm::dyn_cast<MMAAttr>(mmaKind)) {
+    return mmaAttr.getCSingleSubgroupLayout();
+  } else if (auto vmmaAttr = llvm::dyn_cast<VirtualMMAAttr>(mmaKind)) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: also here

**Review by kuhar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 4 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: done!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 10 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: oh, I have always thought llvm has a special dyn_cast, is llvm::dyn_cast and dyn_cast the same?

Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: if it is inded the same, is there a reason why most of the code in IREE is `llvm::dyn_cast` vs regular `dyn_cast`? do we have non uniform name space?

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 8 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: haven't seen this doc, very cool! will do this, thanks!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td (hunk 1 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td]: good point! I didn't remember it was preexisting, but it is. I still think we should, and it's worth the effort, to change existing things that are `InterfaceMethod` to be plain C++ instead.

**Review by bjacob — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 10 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: I think these come from some automatic refactoring script that we ran when migrating from `X.dyn_cast<T>()` to `dyn_cast<T>(X)`

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: but there should be no reason to add `llvm::` in handwritten code

Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: > oh, I have always thought llvm has a special dyn_cast, is llvm::dyn_cast and dyn_cast the same?

What else than `llvm::dyn_cast` could `dyn_cast` be?  (What other namespace has a `dyn_cast` identifier?)

Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: done! although, if you have answers to these Qs would still love to hear it! :)

EDIT: sorry, your previous responses didn't refresh/show up on my screen

Comment (raikonenfnu) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: > What else than llvm::dyn_cast could dyn_cast be? (What other namespace has a dyn_cast identifier?)

Ah, my bad! I was thinking of std C++'s `dynamic_cast`
